### PR TITLE
Add Alert Group demo

### LINKED ISSUE #7008: Alert group demo with overflow
Want a demo that shows an alert group with overflow with a link to open a notification drawer for all notifications. It should match the [designs here](https://docs.google.com/document/d/1sW3b6wJNDCeZaxNQv895APoqZROz6BPRyBNj2D4Rmco/edit?usp=sharing) and use the existing Notification drawer component. See additional notes in the design document.

### PR DESCRIPTION
**Pull request by tompsota**
**What**: Closes #7008

### CODE REVIEW

**Comment by wise-king-sullyman** [comment]
I believe that the main issue here is actually rooted in `AlertGroupInline` using the `index` of each child as the `key` for the list items it creates. Changing the `key` to be based on the `title` of the `Alert` (and adding a `key` to each `Alert` created in the example) rather than the index seems to make things timeout properly in my testing.

The problem with that specific approach is that the `title` isn't going to necessarily be a string since it's typed as a ReactNode, though I'm not sure yet would be more appropriate.

**Comment by tompsota** [comment]
> import React from 'react';
> import { Alert, AlertActionLink, AlertGroup, Button } from '@patternfly/react-core';
> 
> const AlertTimeout: React.FunctionComponent = () => {
>   const [alerts, setAlerts] = React.useState<React.ReactNode[]>([]);
>   const [count, setCount] = React.useState(0);
>   const onClick = () => {
>     const timeout = 3000;
>     setAlerts(prevAlerts => {
>       return [
>         <Alert title={` Alert no. ${count}`} timeout={timeout} actionLinks={
>           <React.Fragment>
>             <AlertActionLink>View details</AlertActionLink>
>             <AlertActionLink>Ignore</AlertActionLink>
>           </React.Fragment>
>         }>
>           This alert will dismiss after {`${timeout / 1000} seconds`}
>         </Alert>,
>         ...prevAlerts,
>       ]
>     });
>   setCount(count+1)
>   }
> 
>   return (
>     <React.Fragment>
>       <Button variant="secondary" onClick={onClick}>Add alert</Button>
>       <Button variant="secondary" onClick={() => setAlerts([])}>Remove all alerts</Button>
>       <AlertGroup isLiveRegion>
>         {alerts}
>       </AlertGroup>
>     </React.Fragment>
>   );
> };

Thanks! This was definitely one of the reasons. Maybe it will be worth creating a separate issue where we could discuss what could be used instead of `index`. I can do that tomorrow. 

If you are curious, the problem was a bit trickier in the demo, because I was tying to remove alert with `onTimeout` prop (which was already done when timeout was over, causing all other alerts to disappear).

**Comment by tompsota** [comment]
> I believe that the main issue here is actually rooted in `AlertGroupInline` using the `index` of each child as the `key` for the list items it creates. Changing the `key` to be based on the `title` of the `Alert` (and adding a `key` to each `Alert` created in the example) rather than the index seems to make things timeout properly in my testing.
> 
> The problem with that specific approach is that the `title` isn't going to necessarily be a string since it's typed as a ReactNode, though I'm not sure yet what would be more appropriate.

Thanks! This was definitely one of the reasons. Maybe it will be worth creating a separate issue where we could discuss what could be used instead of `index`. I can do that tomorrow. 

If you are curious, the problem was a bit trickier in the demo, because I was tying to remove alert with `onTimeout` prop (which was already done when timeout was over, causing all other alerts to disappear).

**Comment by wise-king-sullyman** [comment]
I went ahead and made an issue for the index bug #7877 and put up a PR #7878 (draft for now because I want to add a test when I get a moment).

I _may_ not be understanding your issue properly.

I pulled your branch and as with my smaller scale testing fixing the `index` issue (and making a couple of corresponding changes in your demo, you can see them [on this branch](https://github.com/patternfly/patternfly-react/blob/205b7a62adbe9ff180749b015d1d204365364ff0/packages/react-core/src/demos/AlertGroup/AlertGroupDemo.tsx)) seems to fix the issue of the alerts timing out in the wrong order / at the wrong times.

If you could try making the same changes I did and let me know what the others issue are that I'm missing I can take a look into them 🙂

**Comment by tompsota** [comment]
Soo I compared your branch with mine and found out that the other problem was in `removeAlert` function in `AlertGroupDemo.tsx`. When I changed that function in your code to following (this is the approach I originally used):

```
  const removeAlert = (key: React.Key) => {
    setAlerts(alerts.filter(alert => alert.props.uniqueId !== key.toString()));
  };
```

then all alerts were dismissed at once. Just me being curious, could you explain why this approach doesn't work? I didn't find any particular reason for that.

**Comment by wise-king-sullyman** [comment]
From my understanding of things, I believe what's happening here is that with your original approach `removeAlert` is getting passed to `Alert` with the value of `alerts` *as it is at the time that the alert is created*, so when the function runs the first time it will still have an `alerts` value of `[]`, which `[].filter` then of course returns an empty array and all the alerts are removed.

When using the functional update form in the `setAlerts` call I *am fairly sure* that React (through some form of React internal magic) actually gets the current value for `alerts` as part of the `setAlerts` call.

I might not be 100% right with this explanation as the react docs aren't super clear about it ([they just say to use the functional update form if your new state is dependent on the current state](https://reactjs.org/docs/hooks-reference.html#functional-updates), but don't say why) and when going through blog posts and SO threads there's lots of conflicting "information" about the topic, but this explanation is what makes the most sense to me based on everything I've read and my own testing 😅

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:245]: Rather than creating an array and joining it as a string, we could set this as just a string of the classes, or utilize our Button component to avoid having to set the classes.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:291]:
```suggestion
          isOpen={openDropdownKey === 'dropdown-toggle-id-0'}
```

Currently the kebab dropdown for the notification drawer itself doesn't open.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:344]: Sentence case for "notification drawer"

Additionally:

-  this might be more of a nitpick so let me know what you think, it might  be better to remove the `<br/>` tags and have the instructions wrap normally.
- Could we also add in some verbiage about how the alerts will automatically time out here as well as in the demo description in the MD file?

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:247]: While this styling does work when the viewport width is larger, once it shrinks down the buttons stack on top of one another. So it would be worth adding some vertical margin as well.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:398]:
```suggestion
          <Text component="h2">Max displayed alerts</Text>
```

Since the previous heading used was an `h1`, we should use an `h2` here to avoid skipping heading levels.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:399]:
```suggestion
          <Text component="p">The maximum number of displayed alerts can be set below.</Text>
```

File: packages/react-core/src/demos/AlertGroup.md
Comment (thatblindgeye) [inline packages/react-core/src/demos/AlertGroup.md:17]: This should also be sentence case.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:46]: In case we add more alert group demos down the line, maybe we can update this exported name (plus file name) to something along the lines of `AlertGroupToastWithNotificationDrawer`? What do you think @tompsota ?

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:282]: Can we pass in the `onClose` prop to NotificationDrawerHeader so that a close button for the drawer gets rendered? Right now in order to close the drawer via keyboard, a user would have to Shift + Tab back through the entire page to the bell icon in the header.

**Review by thatblindgeye — CHANGES_REQUESTED** [review]
This is looking great so far! 🎉 I had some comments below (some a little more nitpicky than others, so let me know what you think), but awesome job working on this!

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (wise-king-sullyman) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:135]: Nit: I would swap these and inverse the conditional as I find it's generally a little bit more straightforward for conditionals to not be inverted when it can be avoided.
```suggestion
    if (isActive) {
      setOpenDropdownKey(id);
      return;
    }
    setOpenDropdownKey(null);
```

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (wise-king-sullyman) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:180]: Nit: Rather than having each of these `addXAlertNotification` functions pass so much similar data, what would you think about refactoring `handleNewNotification` to just accept the variant type for the notification and handling the rest of the notification formatting internally?

**Review by wise-king-sullyman — COMMENTED** [review]
I think this is looking really great! I just have a couple of really small nits and one comment that would help DRY things up a touch 🙂

**Comment by tompsota** [comment]
I noticed that the combination of Alert Group with overflow and having the most recent alert on top causes the timeout of alerts that are 'hidden' to start just after they appear on screen.

We discussed this with Austin and we'd like to hear also the opinion from @mcarrano 

We found two possible solutions:

1.  This behavior is okay, because every alert should be displayed on the screen for _n_ seconds (where n is the length of timeout). So the older alerts will appear just after the most recent ones timeout. Also this solution is the easiest, because it doesn't require any further modifications.
2. The timeout should be active also when the alert is 'hidden'. This solution will most likely require us to handle the timeouts outside the alerts - they will be handled in the demo itself. The issue with this approach would be that we don't use the timeout functionality of the `Alert` but add whole new timeout handling.

Personally, I would prefer the first option. IMO, the second one kind of goes against the purpose of the demo -  it should showcase what can be built with the components. When we add add custom timeout handling, then we would showcase just something tailored to the demo instead of the functionality of the `Alert`.

 
The example below shows that the oldest success alert is pushed out to the overflow and its timeout starts just after it is again displayed on the screen, after other (newer) alerts disappear. The timeout is 7 seconds.

https://user-images.githubusercontent.com/72520867/189605020-87fae751-a681-4ea5-97e8-90c004d04c37.mov

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (tompsota) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx:344]: I added the `<br/>`s because the line itself seemed too long for me, but when I think about it now, maybe it'll be better to remove them because the `<br/>`s might break the text on other screen widths.

**Comment by mcarrano** [comment]
@tompsota @wise-king-sullyman I think the suggestion to leave this as it is now is fine. I admit that it is a bit odd, but in reality I don't think you'd want to use this overflow feature together with alerts that time out. The purpose of this is more to manage a queue of alerts that you want the user to read and dismiss manually. Maybe the demo should be set up that way.

BTW, I don't seem to be seeing this demo when I preview the PR. Shouldn't this be showing on a demo tab?

**Comment by thatblindgeye** [comment]
> BTW, I don't seem to be seeing this demo when I preview the PR. Shouldn't this be showing on a demo tab?

@mcarrano currently the demo is in the "Demos" section of the side-nav.

File: packages/react-core/src/demos/AlertGroup.md
Comment (tompsota) [reply packages/react-core/src/demos/AlertGroup.md:3]: Based on the comment below I assume that the demo would fit better in the `components` section 😄

**Comment by tompsota** [comment]
> > BTW, I don't seem to be seeing this demo when I preview the PR. Shouldn't this be showing on a demo tab?
> 
> @mcarrano currently the demo is in the "Demos" section of the side-nav.

Moved the demo to the demo tab

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx
Comment (tompsota) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupDemo.tsx]: That would look better definitely :)

File: packages/react-core/src/demos/AlertGroup.md
Comment (thatblindgeye) [inline packages/react-core/src/demos/AlertGroup.md:15]: This could be moved below the example title heading.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (thatblindgeye) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:240]: @mcoker currently the last alert in the NotificationDrawerBody causes and overflow when its kebab toggle is opened, even if there's enough vertical room for the kebab dropdown:

alert group demo notification drawer overflow

Either disabled the `overflow-y: auto` property or adding a height property to `pf-c-notification-drawer__body` resolves it (I only added a `height: 100%` to the element to test quickly). I haven't tested with any flex properties , though (since the body element is in the `pf-c-notification-drawer` flex container).

Do you think this would be an update that could be made in Core, or maybe just a style specific to this demo?

**Review by thatblindgeye — COMMENTED** [review]
Looking great! I had a couple additional comments below (one nit that I won't block over and one that might be more a CSS issue). I think I'd agree with Matt's comment above about the demo showing alerts that are dismissed manually rather than via timeout, but that isn't really a blocker for me.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (mcoker) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:240]: @thatblindgeye nice catch! That looks like a bug in core to me - https://github.com/patternfly/patternfly/issues/5091

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (mcoker) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:294]: Just curious why you don't use `<Button />` here instead of adding the classes manually?

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (mcoker) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:196]: there is a third `attention` state that we should show if there is a danger alert. I know this is a demo of the alert group, but I think if we're going to show the notification badge based on stuff in the notification drawer, we should make sure this is consistent with our best practices otherwise it could be confusing.

https://github.com/patternfly/patternfly-react/blob/c092c7cfaf877c5783c9ad9db0a8010b8ea023eb/packages/react-core/src/components/NotificationBadge/NotificationBadge.tsx#L11

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (mcoker) [inline packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:215]: FWIW the existing demos are marked as read on click/enter - should this match?

File: packages/react-core/src/demos/AlertGroup.md
Comment (mcoker) [inline packages/react-core/src/demos/AlertGroup.md:10]: This is probably fine, I just wanted to ask if these need to be imported here since they're also imported in [AlertGroupToastWithNotificationDrawer.tsx](https://github.com/patternfly/patternfly-react/blob/69e08e8b523b5fe28dcf817f7f9ca8e982cf6caf/packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx)?

**Review by mcoker — APPROVED** [review]
LGTM! Left some comments, but I don't think any are blockers.

**Comment by tompsota** [comment]
> I think I'd agree with Matt's comment above about the demo showing alerts that are dismissed manually rather than via timeout, but that isn't really a blocker for me.

@thatblindgeye I don't have any problem with removing timeout from the alerts. When I was implementing this demo, I followed the design doc, which mentions that the alerts should have the timeout set to 8 seconds. [Here is the design doc], it is mentioned in the 'Behavior' section (https://docs.google.com/document/d/1sW3b6wJNDCeZaxNQv895APoqZROz6BPRyBNj2D4Rmco/edit#heading=h.ax48avrne16u)

**Review by wise-king-sullyman — APPROVED** [review]
Awesome work on this! 🔥

**Review by thatblindgeye — APPROVED** [review]
@tompsota regarding your latest reply, since it follows the design guidelines I'm fine with it as-is, and would defer to Matt/design regarding whether the demo should have timeouts removed as part of this PR or if it's something to circle back to at some point.

Echoing above, awesome work! 🎉

**Comment by tompsota** [comment]
Please don't merge this PR yet, I'd like to address also comments from @ mcoker. I'll delete this comment when done.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (tompsota) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:276]: @mcoker That definitely is possible and I agree that it would be more consistent.. However, I cannot think about any neat solution. We could access the `NotificationDrawerListItem`s through children of `notificationDrawer` and extract the required data from there, but that approach requires to go quite deep into the tree and there won't be any guarantee that the required props (`variant`, `isRead`) are set. I'd suggest to leave this as it is for now and maybe open this in a new issue, @thatblindgeye what do you think?

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (tompsota) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:215]: From the design doc:
"The user can individually select “Mark as read” or “Clear” on the new notifications from the alert group if they like."
[https://docs.google.com/document/d/1sW3b6wJNDCeZaxNQv895APoqZROz6BPRyBNj2D4Rmco/edit#heading=h.802igegr9dp3](https://docs.google.com/document/d/1sW3b6wJNDCeZaxNQv895APoqZROz6BPRyBNj2D4Rmco/edit#heading=h.802igegr9dp3)
It didn't make sense to me to use both on click/enter and Kebab, so I used only the second one

File: packages/react-core/src/demos/AlertGroup.md
Comment (tompsota) [reply packages/react-core/src/demos/AlertGroup.md:10]: They need to be imported also here, we would get `ReferenceError` otherwise

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (tompsota) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:294]: That's a good point, changed to `<Button>`

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (tompsota) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:196]: I added also this option, I didn't know we have also `attention` state 😅 Thanks!

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (thatblindgeye) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:289]: I haven't yet tried looking into it more closely, but I would agree that this would make sense as a follow-up issue.

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (mcarrano) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:226]: The design guidelines here do specify that the user can make an alert as read either by clicking the list entry or from the kabob menu: https://www.patternfly.org/v4/components/notification-drawer/design-guidelines While it is redundant, I think the idea was to provide a simple one-click way to mark items as read. @tompsota if you can update this, then great. But I also don't feel like it's critical since this demo is not really about demonstrating the notification drawer.

**Review by mcoker — APPROVED** [review]
LGTM!

File: packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx
Comment (tompsota) [reply packages/react-core/src/demos/examples/AlertGroup/AlertGroupToastWithNotificationDrawer.tsx:226]: Updated

**Review by mcarrano — APPROVED** [review]
Thanks for making the updates @tompsota . Looks great!

**Review by tlabaj — COMMENTED** [review]
LGTM! great work.

**Pull request merged**
